### PR TITLE
Clarify named capture group order

### LINKED ISSUE #33333: Missing capture group in table
I think that the table showing the capture groups in the named matched subexpressions section is incorrect.  https://learn.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#named-matched-subexpressions

The table is missing the default named group Two which should be at Number 2.  

In the document:

Number | Name | Pattern
0 | 0 (default name) | ((?\<One\>abc)\d+)?(?\<Two\>xyz)(.\*)
1 | 1 (default name) | ((?\<One\>abc)\d+)
2 | 2 (default name) | (.\*)
3 | One | (?\<One\>abc)
4 | Two | (?\<Two\>xyz)

Should be:

Number | Name | Pattern
0 | 0 (default name) | ((?\<One\>abc)\d+)?(?\<Two\>xyz)(.\*)
1 | 1 (default name) | ((?\<One\>abc)\d+)
2 | 2 (default name) | (?\<Two\>xyz)
3 | 3 (default name) | (.\*)
4 | One | (?\<One\>abc)
5 | Two | (?\<Two\>xyz)

#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 5d124987-d837-b655-261e-e0871effcb7d
* Version Independent ID: 492a3233-0049-0678-1a4e-aed850e14bf4
* Content: [Grouping Constructs in Regular Expressions](https://learn.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions)
* Content Source: [docs/standard/base-types/grouping-constructs-in-regular-expressions.md](https://github.com/dotnet/docs/blob/main/docs/standard/base-types/grouping-constructs-in-regular-expressions.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @adegeo
* Microsoft Alias: **adegeo**

---
[Associated WorkItem - 82745](https://dev.azure.com/msft-skilling/Content/_workitems/edit/82745)

### PR DESCRIPTION
**Pull request by gewarren**
Clarify that named groups come last, and other miscellaneous cleanup.

[*Hide whitespace changes in diff.*](https://github.com/dotnet/docs/pull/34892/files?diff=unified&w=1)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/standard/base-types/grouping-constructs-in-regular-expressions.md](https://github.com/dotnet/docs/blob/9bd51c0ec4a931e42bf6411d0cf1570b891a31d4/docs/standard/base-types/grouping-constructs-in-regular-expressions.md) | [Grouping Constructs in Regular Expressions](https://review.learn.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions?branch=pr-en-us-34892) |

### CODE REVIEW

File: docs/standard/base-types/grouping-constructs-in-regular-expressions.md [hunk 1]
```diff
@@ -14,461 +14,434 @@ helpviewer_keywords:
...
+
+A simple regular expression pattern illustrates how numbered (unnamed) and named groups can be referenced either programmatically or by using regular expression language syntax. The regular expression `((?<One>abc)\d+)?(?<Two>xyz)(.*)` produces the following capturing groups by number and by name. The first capturing group (number 0) always refers to the entire pattern.
+
+| Number | Name             | Pattern                            |
+|--------|------------------|------------------------------------|
+| 0      | 0 (default name) | `((?<One>abc)\d+)?(?<Two>xyz)(.*)` |
+| 1      | 1 (default name) | `((?<One>abc)\d+)`                 |
+| 2      | One              | `(?<One>abc)`                      |
+| 3      | Two              | `(?<Two>xyz)`                      |
+| 4      | 4 (default name) | `(.*)`                             |
```
Comment (gewarren) [inline docs/standard/base-types/grouping-constructs-in-regular-expressions.md:119]: This table is the only place something significant was changed in the file (i.e. the order of the capturing groups).

File: docs/standard/base-types/grouping-constructs-in-regular-expressions.md [hunk 2]
```diff
@@ -14,461 +14,434 @@ helpviewer_keywords:
...
```
Comment (adegeo) [inline docs/standard/base-types/grouping-constructs-in-regular-expressions.md:119]: I tested this and I think the original doc is correct. The suggestion the issue is making seems wrong, it's listing `(?<Two>xyz)` twice in the results.

```csharp
var results = Regex.Match("abc8xyzfdsa", "((?<One>abc)\\d+)?(?<Two>xyz)(.*)");

int index = 0;

//foreach (var group in results.Groups.Cast<Group>()) // .NET 4.8
foreach (var group in results.Groups) // .NET 7
{
    Console.WriteLine($"I:{index} N:{group.Name} V:{group.Value}");
    index++;
}
```

**.NET 7**
```output
I:0 N:0 V:abc8xyzfdsa
I:1 N:1 V:abc8
I:2 N:2 V:fdsa
I:3 N:One V:abc
I:4 N:Two V:xyz
```

**.NET Framework** (same as .NET 7 but I still included it)
```output
I:0 N:0 V:abc8xyzfdsa
I:1 N:1 V:abc8
I:2 N:2 V:fdsa
I:3 N:One V:abc
I:4 N:Two V:xyz
```

**Review by adegeo — CHANGES_REQUESTED** [review]

File: docs/standard/base-types/grouping-constructs-in-regular-expressions.md [hunk 3]
```diff
@@ -14,461 +14,434 @@ helpviewer_keywords:
...
+
+For information on groups and the regular expression object model, see [Grouping constructs and regular expression objects](#grouping-constructs-and-regular-expression-objects).
+
+## Matched subexpressions
+
+The following grouping construct captures a matched subexpression:
+
+`(` *subexpression* `)`
+
+Here, *subexpression* is any valid regular expression pattern. Captures that use parentheses are numbered automatically from left to right based on the order of the opening parentheses in the regular expression, starting from 1. However, [named capture groups](#named-matched-expressions) are always ordered last, after non-named capture groups. The capture that's numbered 0 is the text matched by the entire regular expression pattern.
```
Comment (gewarren) [inline docs/standard/base-types/grouping-constructs-in-regular-expressions.md:49]:
```suggestion
Here, *subexpression* is any valid regular expression pattern. Captures that use parentheses are numbered automatically from left to right based on the order of the opening parentheses in the regular expression, starting from 1. However, [named capture groups](#named-matched-subexpressions) are always ordered last, after non-named capture groups. The capture that's numbered 0 is the text matched by the entire regular expression pattern.
```

**Review by adegeo — APPROVED** [review]
LGTM!

**Pull request merged**